packages/cw721/src/msg.rs

```diff
@@ -21,15 +49,624 @@ pub enum Cw721ExecuteMsg {
+                        // None: no royalty info, so create new
+                        None => {
+                            updated.royalty_info =
+                                Some(royalty_info_response.create(deps, env, info, None)?);
+                        }
+                    }
+                }
+                Ok(updated)
+            }
+            // None: create new metadata
+            None => {
+                let royalty_info = match &self.royalty_info {
+                    // new royalty info
+                    Some(royalty_info) => Some(royalty_info.create(deps, env, info, None)?),
+                    // current royalty is none and new royalty is none
+                    None => None,
+                };
+                let new = CollectionMetadataExtension {
+                    description: self.description.clone().unwrap_or_default(),
+                    image: self.image.clone().unwrap_or_default(),
+                    external_link: self.external_link.clone(),
+                    explicit_content: self.explicit_content,
+                    start_trading_time: self.start_trading_time,
+                    royalty_info,
+                };
+                Ok(new)
+            }
+        }
+    }
+
+    fn validate(
+        &self,
+        deps: Deps,
+        _env: &Env,
+        info: &MessageInfo,
+        _current: Option<&CollectionMetadataExtension<RoyaltyInfo>>,
+    ) -> Result<(), Cw721ContractError> {
+        // start trading time can only be updated by minter
+        let minter_initialized = MINTER.item.may_load(deps.storage)?;
+        if self.start_trading_time.is_some()
+            && minter_initialized.is_some()
+            && MINTER.assert_owner(deps.storage, &info.sender).is_err()
+        {
+            return Err(Cw721ContractError::NotMinter {});
+        }
+        // all other props collection metadata extension can only be updated by the creator
+        let creator_initialized = CREATOR.item.may_load(deps.storage)?;
+        if (self.description.is_some()
+            || self.image.is_some()
+            || self.external_link.is_some()
+            || self.explicit_content.is_some()
+            || self.royalty_info.is_some())
+            && creator_initialized.is_some()
+            && CREATOR.assert_owner(deps.storage, &info.sender).is_err()
+        {
+            return Err(Cw721ContractError::NotCollectionCreator {});
+        }
+        // check description length, must not be empty and max 512 chars
+        if let Some(description) = &self.description {
+            if description.is_empty() {
+                return Err(Cw721ContractError::CollectionDescriptionEmpty {});
+            }
+            if description.len() > MAX_COLLECTION_DESCRIPTION_LENGTH as usize {
+                return Err(Cw721ContractError::CollectionDescriptionTooLong {
+                    max_length: MAX_COLLECTION_DESCRIPTION_LENGTH,
+                });
+            }
+        }
```

Comment:
Pls also note that `StateFactory` can also do complex validation. Like here in `CollectionMetadataExtension`:
- `start_trading_time` can only be updated by Minter
- all other probs belongs to and can only be updated by Creator :sunglasses: 

In short: all state-based logic can be handled in StateFactory.

@jhernandezb @humanalgorithm @JakeHartnell